updater.go

```diff
@@ -244,19 +245,108 @@ func (u *Updater) AppendHeaderAt(fh *FileHeader, offset int64) (io.Writer, error
 		return nil, err
 	}
 
-	if offset < 0 {
-		offset = u.dirOffset
-	}
 	// Offset should match existing header offsets or equal to directory offset.
 	var offsetExists bool
-	for i, h := range u.dir {
-		if h.offset == uint64(offset) {
-			offsetExists = true
-			// Delete the corresponding header.
-			u.dir = append(u.dir[:i], u.dir[i+1:]...)
-			break
+
+	existingFileIndex := 0
+	if offset < 0 {
+		// Append the file to the zip or replace the existing one
+		for i, h := range u.dir {
+			if h.FileHeader.Name == fh.Name {
+				offset = int64(h.offset)
+				existingFileIndex = i
+				break
+			}
+		}
+
+		// File does not exist inside the zip, so we append it
+		if existingFileIndex == 0 {
+			offset = u.dirOffset
+		}
+	} else {
+		// This is for manually writing data into a certain offset
+		for i, h := range u.dir {
+			if h.offset == uint64(offset) {
+				offsetExists = true
+				// Delete the corresponding header.
+				u.dir = append(u.dir[:i], u.dir[i+1:]...)
+				break
+			}
 		}
 	}
+
+	if existingFileIndex > 0 {
+		// If the file exists and is not the last one in the file, we will remove it
+		// and reinsert it at the end of the file.
+		// The existing files data will be relocated first to use the leftover space.
+
+		// This diagram shows the internal structure of a zip file, whose i-th file
+		// we want to overwrite. The annotations represent some of the variables that
+		// are used to implement this feature.
+		//
+		//                       deletedDataSize = nextOffset - existingFileOffset
+		//                      ┌────────────────┐
+		//                      │                │ remainingDataSize
+		//                      │                │────────────────────┐
+		//                      ▼                ▼                    ▼
+		// ┌───┬────────────┬───┬───┬────────────┬───┬────────────┬───┬───────┐
+		// │   │            │   │   │            │   │            │   │░░░░░░░│
+		// │h_0│   File 0   │...│h_i│   File i   │h_ │  File i+1  │...│░░Dir░░│
+		// │   │    data    │   │   │    data    │i+1│    data    │   │░░░░░░░│
+		// │   │            │   │   │            │   │            │   │░░░░░░░│
+		// └───┴────────────┴───┴───┴────────────┴───┴────────────┴───┴───────┘
+		//                      ▲                ▲                    ▲
+		//                      │                │                    │
+		//                   existingFileOffset  nextOffset           Dir Offset
+		if existingFileIndex+1 < len(u.dir) {
+			existingFileOffset := u.dir[existingFileIndex].offset
+			nextOffset := u.dir[existingFileIndex+1].offset
+
+			// Read the existing files data
+			remainingDataSize := u.dirOffset - int64(nextOffset)
+			data := make([]byte, remainingDataSize)
```

Comment:
I'm concerned about replacing files in the large zip file. Should there be a safer way to allocate a small buffer and rewind existing file data in sections instead of allocating the entire memory?

Reply:
Yeah, that should probably be the way to do it properly.

Reply:
I've changed the code to rewind existing file data on a file by file basis. Let me know what you think. It could still potentially be done using smaller chunks instead of moving whole files.

Reply:
Here are some of my perspectives:
- If the zip archive stores many large files (10GB for example), rewinding file by file will still allocate large memory many times (or OOM killed by system directly).
- If the zip archive stores many small files (10Bytes for example), rewinding file by file will cause a poor performance since it needs to read & write in small data size many times.

These scenarios although uncommon but still possible. So from my pov, we need to create a small buffer (maybe 10K for example, or make it dynamic based on the remaining data size), and the file rewinding logic can be moved to a separate function if it needs many logics.
What do you think of it? :-)